pyproject.toml

```diff
@@ -6,113 +6,298 @@ requires = [
+authors = [{name = "The SunPy Community", email = "sunpy@googlegroups.com"}]
+license = {text = "BSD 2-Clause License"}
+keywords = ["solar physics", "solar", "science", "sun", "wcs", "coordinates"]
+classifiers = [
+  "Development Status :: 5 - Production/Stable",
+  "Intended Audience :: Science/Research",
+  "License :: OSI Approved :: BSD License",
+  "Natural Language :: English",
+  "Operating System :: OS Independent",
+  "Programming Language :: Python",
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3.9",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
+  "Topic :: Scientific/Engineering :: Physics",
+]
+requires-python = ">=3.9"
+dependencies = [
+  # Exclude versions of astropy which are incompatible with matplotlib 3.7
+  "astropy>=5.0.6,!=5.1.0",
+  "numpy>=1.21.0",
+  "packaging>=19.0",
+  "parfive[ftp]>=2.0.0",
+  "requests",
+]
 
-# Until we have h5py binaries we can't test wheels on Python 3.11
-[tool.cibuildwheel]
-test-command = "pytest -p no:warnings --doctest-rst -m 'not mpl_image_compare' --pyargs sunpy"
-test-extras = ["tests"]
+[project.readme]
+file = "README.rst"
+content-type = "text/x-rst"
 
-[ tool.gilesbot ]
-  [ tool.gilesbot.circleci_artifacts ]
-    enabled = true
+[project.urls]
+Homepage = "https://sunpy.org"
+Download = "https://pypi.org/project/sunpy/"
+"Source Code" = "https://github.com/sunpy/sunpy/"
+Documentation = "https://docs.sunpy.org/"
+Changelog = "https://docs.sunpy.org/en/stable/whatsnew/changelog.html"
+"Issue Tracker" = "https://github.com/sunpy/sunpy/issues"
 
-  [ tool.gilesbot.circleci_artifacts.figure_report ]
-    url = ".tmp/py310-figure/figure_test_images/fig_comparison.html"
-    message = "Click details to see the figure test comparisons, for py310-figure."
-    report_on_fail = true
+[project.optional-dependencies]
+asdf = [
+  "asdf-astropy>=0.1.1",
+  "asdf>=2.8.0",
+]
+dask = ["dask[array]>=2021.4.0"]
+image = [
+  "scikit-image>=0.18.0",
+  # !=1.10.0 due to https://github.com/scipy/scipy/issues/17718
+  "scipy>=1.7.0,!=1.10.0",
+]
+jpeg2000 = [
+  "glymur>=0.9.1,!=0.9.5",
+  "lxml>=4.8.0,!=5.0.0",
+]
+map = [
+  "matplotlib>=3.5.0",
+  "mpl-animators>=1.0.0",
+  "reproject",
+  # !=1.10.0 due to https://github.com/scipy/scipy/issues/17718
+  "scipy>=1.7.0,!=1.10.0",
+]
+net = [
+  "beautifulsoup4>=4.8.0",
+  "drms>=0.6.1,<0.7.0",
+  "python-dateutil>=2.8.0",
+  "tqdm>=4.32.1",
+  "zeep>=3.4.0",
+]
+spice = ["spiceypy"]
+timeseries = [
+  "cdflib>=0.3.20,!=0.4.0,!=1.0.0",
+  "h5netcdf>=0.11",
+  # While a not direct dependency
+  # We need to raise this to ensure the goes netcdf files open.
+  "h5py>=3.1.0",
+  "matplotlib>=3.5.0",
+  "pandas>=1.2.0",
+]
+visualization = [
+  "matplotlib>=3.5.0",
+  "mpl-animators>=1.0.0",
+]
+all = [
+  "sunpy[asdf,dask,image,jpeg2000,map,net,spice,timeseries,visualization]",
```

Comment:
This syntax didn't used to be supported right? What version of setuptools etc do we need to support this?

Reply:
I will have to check and increase the min version.

Reply:
I worry that if it's a pip version we wont be able to use it for a while. We should work on the principle that the min pip version is the one that ships with the min version of Python we have.

Reply:
That being said I suspect it's setuptools not pip.

Reply:
I know I had someone not able to instal a project due to an old pip that only had a pyproject. I forget if their update of pip also brought in setuptools (i would assume it had it)

Reply:
I pulled the 3.9.0 docker container.

```
Python 3.9.0 (default, Nov 18 2020, 13:28:38) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pip
>>> pip.__version__
'20.3.1'
>>> import setuptools
>>> setuptools.__version__
'50.3.2'
```

As long as the build dependencies are installed correctly, I don't *think* we need to worry about the setuptools version.

Doing this `pip install git+https://github.com/nabobalis/sunpy@pyproj#egg=sunpy[all]` inside that container, did not work. It started looking for sunpy releases on pypi with all the extras we needed.

Doing it on python 3.10.0 worked though, which would be our minimum Python for the next release? So I think this PR is ok as long as we bump to min 3.10 at the same time.

Reply:
I will bump